### DISCUSSION TITLE
RDF Syntax Test Suites

### DISCUSSION THREAD

**Discussion by namedgraph** [RDFLib/rdflib#2437] [General]
I see that rdflib is passing the [RDF Core Test Results](https://www.w3.org/2003/11/results/rdf-core-tests). What about [RDF Syntax Test Suites (Turtle family)](https://www.w3.org/2011/rdf-wg/wiki/RDF_Test_Suites#RDF_Syntax_Test_Suites_.28Turtle_family.29) though? I think they should be able to catch some of the `Dataset`-related problems.

**Comment by namedgraph** [comment]
OK it looks like the W3C edit included: https://github.com/RDFLib/rdflib/tree/main/test/data/suites/w3c
I don't get how they are passing though?

**Comment by ghost** [comment]
> I think they should be able to catch some of the `Dataset`-related problems.
> I don't get how they are passing though?

The issues pertaining to the RDFLib Dataset implementation aren't related to the parsing of RDF.

**Reply by aucampia** [reply]
Also, the syntax tests are rather superficial. There is some infrastructure for round trip testing, but I think there we also get lucky. We also have some infrastructure for testing various variants of the same graph, and I'm extending this to also allow one variant to be python code. This will make it easier to confirm that various syntaxes results in the right thing, but there are still some issues we will miss.

Right now I'm trying to fix Literal datatype and lang strings (<https://github.com/RDFLib/rdflib/issues/2123>, <https://github.com/RDFLib/rdflib/issues/1326>, <https://github.com/RDFLib/rdflib/issues/670>), as it is causing some trouble with other tests.

**Reply by namedgraph** [reply]
OK my bad, then syntax tests are not the right ones. But simply round-tripping test data through `Dataset.parse`/`Dataset.serialize` and comparing the input with the output should highlight `Dataset` the issues then? Are there tests like that? If not, how/where can they be contributed?

**Reply by aucampia** [reply]
We have round trip tests here:

https://github.com/RDFLib/rdflib/blob/4b96e9d18cf42b64237d232fcb88cab5b316e877/test/test_roundtrip.py

I think they have picked up some issues, we have many xfail tests [[ref](https://github.com/search?q=repo%3ARDFLib%2Frdflib%20xfail%20path%3Atest%2F&type=code)]. But it is also worth considering that because the publicID was being set for them, and because this was used for default graph name, this somewhat masked some problems. The fix for default graph name only eliminated one xfail test.

I think publicID should be set to correctly handle relative URIs, though I could be wrong about this.

The round trip tests are very similar to the variant tests:

https://github.com/RDFLib/rdflib/blob/4b96e9d18cf42b64237d232fcb88cab5b316e877/test/test_graph/test_variants.py

These compare the different variants of what should be the same graph to ensure they are identical, I'm busy expanding and improving this, making the test more fine-grained and also adding support for python variants, but this is somewhat blocked by problems with Literals.

**Reply by aucampia** [reply]
The default graph name issue was one of the biggest with Dataset I think. There are other problems, but I think they are mostly related to the weirdness with the context being a `Graph` object instead of the graph name in most place, and the store interface is really over-complicated, but for those we really need very specific issues detailing exact problems to understand where what is going wrong. We can also try to redesign a store interface from scratch, and we certainly should consider it, but it is still worth understanding where things goes wrong.

**Reply by namedgraph** [reply]
What is the `publicID` for?

**Reply by ghost** [reply]
> the store interface is really over-complicated

If you'll forgive my curiosity, I'd be grateful if could you elaborate slightly, just a single example would suffice to help me understand.

**Reply by ghost** [reply]
> What is the `publicID` for?

It's for [multi-graph sources](https://rdflib.readthedocs.io/en/stable/plugin_parsers.html#multi-graph-ids)

**Reply by aucampia** [reply]
> > the store interface is really over-complicated
>
> If you'll forgive my curiosity, I'd be grateful if could you elaborate slightly, just a single example would suffice to help me understand.

- The interface takes a Graph object as Context, which means that you have to pass in a graph (a container for triples) when asking for the triples inside the graph. Graph is a mutable object, and this is using a mutable object is used as an identifier. This to me is incredibly confusing.
- It includes namespace management, which conflates two separate concerns which could be provided by two different interfaces, the code for the namespace management also suggests it is optional, which I think it is not, and if it is not implemented there is no fallback logic, and things like turtle serialization won't work as it calls computer_qname which calls into the namespace management which calls the store. The whole approach with namespace management where it strings from namespace manager to graph to store is not needed, there can be a namespace manager with a namespace store if we really want to provide it, if a plugin wants to provide both with one object they can, but if a plugin only wants to provide triple storage they can ignore namespace management. This also consequently imports the over complicated namespace management which has led to issues like this <https://github.com/RDFLib/rdflib/issues/2077>.
- Optional methods are not on subclasses making it difficult to understand what a store can do using typing (issubclass, isinstance) and also consequently eliminates any safety that could be provided by type hints requiring programmers to instead keep this complexity in their heads.
- The create and open methods imply that a store could be invalid state, while there are also no methods to test for created or opened state. A simpler option would have been to have these as static methods or on a factory [[ref](https://arne-mertz.de/2016/01/avoid-invalid-state/)][[ref](https://fsharpforfunandprofit.com/posts/designing-with-types-making-illegal-states-unrepresentable/)], making it impossible to have a store in an invalid state. 
- The open method uses return values for indicating errors instead of raising exceptions, meaning that programmers have to use a non-idiomatic error handling pattern to implement it.
- There are methods with unclear semantics:
  - `gc` (which I think is not called ever)
  - `add_graph`/`remove_graph` which says that it adds or removes a graph, but a graph is a container of triples, and the implementations does not add all triples (see the first point).
- It contains methods that would preclude implementations that support multithreading, like commit and rollback. A transaction, if it works, should be implemented as a separate object not in the store. This is how almost every database library works.
- Store objects can be context aware or not context aware, and implementors have to understand RDFLib behaviour with regard to this flag to correctly implement the store. A simpler approach would have been to instead use subclassing or just different classes for context aware and non-context aware classes.

I think also that oxrdflib finds it difficult to implement our interface in a way that provides working datasets (<https://github.com/oxigraph/oxrdflib/issues/28>) speak to this, but this of course could also just be a symptom of other problems.

**Reply by aucampia** [reply]
> > What is the `publicID` for?
> 
> It's for [multi-graph sources](https://rdflib.readthedocs.io/en/stable/plugin_parsers.html#multi-graph-ids)

After [#2406: BREAKING CHANGE: Don't use publicID as the name for the default graph.](https://github.com/RDFLib/rdflib/pull/2406) this is no longer the case, but thanks for highlighting it, I will fix that document.

The default graph, according to w3c specs, have no identifier, so having the publicID as it's identifier is clearly not standards compliant, and an issue many users of RDFLib have complained about for years.

After  [#2406: BREAKING CHANGE: Don't use publicID as the name for the default graph.](https://github.com/RDFLib/rdflib/pull/2406) the only use of the publicID is as the base_uri for resolving relative URIs.

https://rdflib.readthedocs.io/en/latest/apidocs/rdflib.html#rdflib.Graph.parse

> publicID ([Optional](https://docs.python.org/3.7/library/typing.html#typing.Optional)[[str](https://docs.python.org/3.7/library/stdtypes.html#str)]) –
> 
> the logical URI to use as the document base. If None specified the document location is used (at least in the case where there is a document location). This is used as the base URI when resolving relative URIs in the source document, as defined in [IETF RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-5.1.4), given the source document does not define a base URI.

Any other purpose for publicID should be documented in the parse method, but for now this is the only defined behaviour, any other use of it is a defect, either with the docs or more likely with the code.

The default graph should be the default graph, and it is still not quite, as someone can create an RDF Dataset that collides with our default graph, but this is a bug because again, the default graph should have no name, not a special reserved name.

**Reply by aucampia** [reply]
@gjhiggins if you or anyone else disagrees and think that the publicID should be used as the default graph ID in addition to the base URI for resolving relative URIs, as it was before https://github.com/RDFLib/rdflib/pull/2406, please open an issue, so we can discuss reverting https://github.com/RDFLib/rdflib/pull/2406. Otherwise, the behaviour you described here was wrong, and was a bug, and should not persist, and was correctly fixed.

**Reply by namedgraph** [reply]
Sounds like `publicID` should be called `baseURI`?

**Reply by aucampia** [reply]
@namedgraph well, `base_uri` :sweat_smile: , but that would be a better name as that is the name used by the IETF spec for URIs.

I made this issue now to clearly track the problem with the default context having a name:

- https://github.com/RDFLib/rdflib/issues/2445

To be clear though, this is not the biggest issue in the world, we can kind of get away with it, but it is an issue.

**Reply by aucampia** [reply]
- https://github.com/RDFLib/rdflib/discussions/2447

**Reply by ghost** [reply]
> @gjhiggins if you or anyone else disagrees and think that the publicID should be used as the default graph ID in addition to the base URI for resolving relative URIs, as it was before #2406, please open an issue, so we can discuss reverting #2406. Otherwise, the behaviour you described here was wrong, and was a bug, and should not persist, and was correctly fixed.

I don't *disagree* as such, I just don't have a particularly detailed model and I'm not going to argue merely for the sake of arguing. You've identified a potential interaction between base URI for resolving relative URIs and the use of publicID for controlling the parsed destination of multi-graph source docs and I haven't had time to investigate and form a personal opinion other than that I can see a user might want to preserve any existing default graph content from "contamination" when parsing a multi-graph source doc and publicID seems to have been the way to do that thus far. But if there are negative consequences of (ab)using publicID (assuming it is an abuse) then avoiding those consequences would take priority.

**Reply by aucampia** [reply]
> a user might want to preserve any existing default graph content from "contamination" when parsing a multi-graph source doc and publicID seems to have been the way to do that thus far.

For all non-default graphs, the behaviour of calling `Dataset.parse()` is that any triples from graph in the source with the same name gets added to the graph. Why should this be different for the default graph? Why for the default graph, instead of merging it, as we do for every other graph, we all of a sudden give it a name in contravention to the spec, and then randomly decide the name should be the same as the base URI for the document?

I don't see how this is possibly a positive consequence, and if someone wants it, they should explain why. I don't know of anyone who wants it, given the amount of issues we have had raised for this behaviour I think it is not really defensible.

If `{Dataset,Graph}.parse` updates the dataset or graph, it should do so consistently. If someone then does not want to update `{Graph,Dataset}.parse`, they should not call parse on a `Graph` or `Dataset` they don't want updated, and instead create a new Graph or Dataset, and call parse on that.

**Reply by aucampia** [reply]
@gjhiggins if the standard is things should behave the way they have behaved before, I don't really think there is much point in us making changes to RDFLib any longer, and it would be better to just say it is what it is and move on or start a different project from scratch.

**Reply by ghost** [reply]
> @gjhiggins if the standard is things should behave the way they have behaved before

No, that's *not* what I believe but I'm obviously miscommunicating to markedly ill effect, to the extent that I'm actually being inadvertently disruptive. Time for me to stfu.

**Comment by namedgraph** [comment]
@gjhiggins what is your point exactly? Are you saying preserving the current behaviour is more important than standards compliance? Because the `Dataset` is not compliant in many ways, and @aucampia has used his time to identify and fix the issues.

Do you agree that rdflib should strive for 100% standards compliance? And that version 7 is a good way to improve it even if it means breaking changes (because they will not impact existing code)? If not, then I think it's best to fork this codebase into a different project that will work towards compliance, and leave rdflib behind as a legacy project. Having multiple projects wouldn't be great for the Python RDF community, but it also deserves to have a compliant library.
